File: cli/src/core/RoutePlan.ts
/**
 * RoutePlan — parses the route LLM's JSON-in-text output into a per-topic
 * assignment map, mapping source ordinals back to SourceRefs. Fail-loud on
 * truncation or malformed JSON (the caller aborts the batch and retries).
 */

import { createLogger } from "../Logger.js";
import type { SourceRef } from "./TopicKBTypes.js";

const log = createLogger("RoutePlan");

export interface TopicAssignment {
	readonly title: string | undefined; // present for new topics
	readonly isNew: boolean;
	readonly refs: ReadonlyArray<SourceRef>;
}

export interface RoutePlan {
	/** stableSlug → assignment. Empty when an error occurred. */
	readonly assignments: Map<string, TopicAssignment>;
	/** Set when parsing failed (truncation / malformed) — caller marks nothing processed. */
	readonly error?: string;
}

interface RawUpdate {
	stableSlug?: unknown;
	title?: unknown;
	sourceIndexes?: unknown;
}

/** Strips an optional ```json … ``` fence the LLM may wrap the object in. */
function stripFence(text: string): string {
	const trimmed = text.trim();
	const fence = trimmed.match(/^```(?:json)?\s*\n([\s\S]*?)\n```$/);
	return fence ? fence[1].trim() : trimmed;
}

export function parseRoutePlan(
	text: string,
	stopReason: string | null | undefined,
	batch: ReadonlyArray<SourceRef>,
): RoutePlan {
	if (stopReason === "max_tokens") {
		return { assignments: new Map(), error: "route output truncated at max_tokens" };
	}
	let raw: { updates?: unknown; newTopics?: unknown };
	try {
		raw = JSON.parse(stripFence(text));
	} catch {
		return { assignments: new Map(), error: "route output is not valid JSON" };
	}

	const assignments = new Map<string, TopicAssignment>();
	// An out-of-range / non-numeric index means the route response cannot be
	// trusted to map ordinals to sources correctly. Silently dropping it would
	// leave the real source the LLM miscounted unreferenced — which the caller
	// then consumes as "un-filed", losing it permanently with no trail. Instead
	// flag the whole response as malformed (fail-loud, same as max_tokens / bad
	// JSON) so the caller holds the batch and retries rather than dropping data.
	let malformedIndex = false;
	const add = (entry: RawUpdate, isNew: boolean): void => {
		if (typeof entry?.stableSlug !== "string" || entry.stableSlug.length === 0) return;
		const indexes = Array.isArray(entry.sourceIndexes) ? entry.sourceIndexes : [];
		const refs: SourceRef[] = [];
		for (const idx of indexes) {
			if (typeof idx !== "number" || idx < 0 || idx >= batch.length) {
				log.warn("route: out-of-range source index %o for topic %s — failing route", idx, entry.stableSlug);
				malformedIndex = true;
				continue;
			}
			refs.push(batch[idx]);
		}
		if (refs.length === 0) return;
		const candidateTitle = isNew && typeof entry.title === "string" ? entry.title : undefined;
		const existing = assignments.get(entry.stableSlug);
		if (existing) {
			// Union-merge: a slug the LLM filed under both `updates` and `newTopics`
			// must keep the "new topic" flag and its title. Spreading `...existing`
			// would pin isNew/title to whichever array was iterated first (updates,
			// which carries neither), so reconcile would treat a brand-new topic as
			// an update and the page title would degrade to the slug.
			// `refs` is ReadonlyArray, never mutated in place — build a new array.
			const merged = [...existing.refs, ...refs.filter((r) => !existing.refs.includes(r))];
			assignments.set(entry.stableSlug, {
				title: existing.title ?? candidateTitle,
				isNew: existing.isNew || isNew,
				refs: merged,
			});
			return;
		}
		assignments.set(entry.stableSlug, { title: candidateTitle, isNew, refs });
	};

	for (const u of Array.isArray(raw.updates) ? raw.updates : []) add(u as RawUpdate, false);
	for (const n of Array.isArray(raw.newTopics) ? raw.newTopics : []) add(n as RawUpdate, true);

	if (malformedIndex) {
		return { assignments: new Map(), error: "route referenced an out-of-range source index" };
	}
	return { assignments };
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/RoutePlan.kt
package ai.jolli.jollimemory.core

import com.google.gson.JsonElement
import com.google.gson.JsonParser

/**
 * RoutePlan — parses the route LLM's JSON-in-text output into a per-topic
 * assignment map, mapping source ordinals back to [SourceRef]s. Fail-loud on
 * truncation / malformed JSON / out-of-range index (caller holds the batch).
 *
 * Kotlin port of `cli/src/core/RoutePlan.ts`.
 */
data class TopicAssignment(
    val title: String?, // present for new topics
    val isNew: Boolean,
    val refs: List<SourceRef>,
)

data class RoutePlan(
    /** stableSlug → assignment. Empty when an error occurred. */
    val assignments: Map<String, TopicAssignment>,
    /** Set when parsing failed (truncation / malformed) — caller marks nothing processed. */
    val error: String? = null,
)

object RoutePlanParser {

    private val log = JmLogger.create("RoutePlan")

    private val FENCE_RE = Regex("^```(?:json)?\\s*\\n([\\s\\S]*?)\\n```$")

    /** Strips an optional ```json … ``` fence the LLM may wrap the object in. */
    private fun stripFence(text: String): String {
        val trimmed = text.trim()
        val fence = FENCE_RE.find(trimmed)
        return fence?.groupValues?.get(1)?.trim() ?: trimmed
    }

    fun parseRoutePlan(text: String, stopReason: String?, batch: List<SourceRef>): RoutePlan {
        if (stopReason == "max_tokens") {
            return RoutePlan(emptyMap(), "route output truncated at max_tokens")
        }
        val root = try {
            JsonParser.parseString(stripFence(text))
        } catch (_: Exception) {
            return RoutePlan(emptyMap(), "route output is not valid JSON")
        }
        if (!root.isJsonObject) return RoutePlan(emptyMap(), "route output is not valid JSON")
        val obj = root.asJsonObject

        val assignments = LinkedHashMap<String, TopicAssignment>()
        var malformedIndex = false

        fun add(entry: JsonElement, isNew: Boolean) {
            if (!entry.isJsonObject) return
            val e = entry.asJsonObject
            val slugEl = e.get("stableSlug")
            val slug = if (slugEl != null && slugEl.isJsonPrimitive && slugEl.asJsonPrimitive.isString) slugEl.asString else ""
            if (slug.isEmpty()) return

            val indexesEl = e.get("sourceIndexes")
            val refs = mutableListOf<SourceRef>()
            if (indexesEl != null && indexesEl.isJsonArray) {
                for (idxEl in indexesEl.asJsonArray) {
                    val isNumber = idxEl.isJsonPrimitive && idxEl.asJsonPrimitive.isNumber
                    val asDouble = if (isNumber) idxEl.asDouble else Double.NaN
                    val idx = asDouble.toInt()
                    // Reject non-numbers, non-integers (e.g. 1.5 — asInt would silently
                    // truncate to a valid-looking ordinal), and out-of-range: a miscounted
                    // index means the ordinal→source mapping can't be trusted, so fail loud.
                    if (!isNumber || asDouble != idx.toDouble() || idx < 0 || idx >= batch.size) {
                        log.warn("route: invalid source index %s for topic %s — failing route", idxEl.toString(), slug)
                        malformedIndex = true
                        continue
                    }
                    refs.add(batch[idx])
                }
            }
            if (refs.isEmpty()) return

            val titleEl = e.get("title")
            val candidateTitle = if (isNew && titleEl != null && titleEl.isJsonPrimitive && titleEl.asJsonPrimitive.isString) titleEl.asString else null

            val existing = assignments[slug]
            if (existing != null) {
                // Union-merge: keep the "new topic" flag + title across both arrays.
                val merged = existing.refs + refs.filter { it !in existing.refs }
                assignments[slug] = TopicAssignment(
                    title = existing.title ?: candidateTitle,
                    isNew = existing.isNew || isNew,
                    refs = merged,
                )
                return
            }
            assignments[slug] = TopicAssignment(title = candidateTitle, isNew = isNew, refs = refs)
        }

        (obj.get("updates")?.takeIf { it.isJsonArray }?.asJsonArray ?: emptyList<JsonElement>()).forEach { add(it, false) }
        (obj.get("newTopics")?.takeIf { it.isJsonArray }?.asJsonArray ?: emptyList<JsonElement>()).forEach { add(it, true) }

        if (malformedIndex) {
            return RoutePlan(emptyMap(), "route referenced an out-of-range source index")
        }
        return RoutePlan(assignments)
    }
}
